File: tools/generate_fixtures.py
"""Fixture loader and inspector for NLvib validation fixtures.

This script serves two purposes:

1. **Library** — provides :func:`load_fixture` for use in tests and notebooks.
2. **CLI tool** — ``python tools/generate_fixtures.py --list`` prints a summary
   table of all fixtures found under ``tests/fixtures/``.

The companion MATLAB script ``tools/generate_fixtures.m`` generates the ``.npz``
files from the original NLvib MATLAB source.  This Python script does not invoke
MATLAB; it only loads and inspects the already-generated fixture files.

Usage::

    # List all fixtures
    python tools/generate_fixtures.py --list

    # Inspect a specific fixture
    python tools/generate_fixtures.py --inspect 01_Duffing

    # Load in Python code
    from tools.generate_fixtures import load_fixture
    fx = load_fixture("01_Duffing")
    print(fx["omega"].shape)

Schema
------
Every fixture is a dict with at minimum:

    omega      : ndarray, shape (n_points,)  — frequency (rad/s)
    amplitude  : ndarray, shape (n_points,)  — response amplitude at DOF 0
    phase      : ndarray, shape (n_points,)  — phase of fundamental harmonic (rad)
    stability  : ndarray, shape (n_points,), dtype bool
    tolerance  : float  — validation tolerance (1e-6 default, 1e-4 for non-smooth)

Additional arrays (omega_shoot, amplitude_dof2, Q_harmonics, …) may be present
depending on the example; see ``tests/fixtures/README.md`` for the full catalogue.
"""

from __future__ import annotations

import argparse
import sys
from pathlib import Path
from typing import Any

import numpy as np

# ---------------------------------------------------------------------------
# Paths
# ---------------------------------------------------------------------------
_REPO_ROOT = Path(__file__).parent.parent
_FIXTURES_DIR = _REPO_ROOT / "tests" / "fixtures"

# Canonical list of all expected MATLAB-generated fixtures (name → description).
_EXPECTED_FIXTURES: dict[str, str] = {
    "01_Duffing": "1-DOF Duffing oscillator — HB + Shooting overlay",
    "02_twoDOFoscillator_cubicSpring": "2-DOF chain with cubic spring — HB",
    "03_twoDOFoscillator_unilateralSpring": "2-DOF chain with unilateral spring — HB",
    "04_twoDOFoscillator_cubicSpring_NM": "2-DOF cubic spring — NMA backbone",
    "05_twoDOFoscillator_tanhDryFriction_NM": "2-DOF tanh friction — NMA backbone",
    "06_twoDOFoscillator_tanhDryFriction_FRF": "2-DOF tanh friction — HB FRF",
    "07_geometricNonlinearity": "Multi-DOF geometric nonlinearity — HB",
    "08_multiDOF_multiNL": "Multi-DOF with multiple NL elements — HB",
    "09_beam_tanhFriction": "FE beam + tanh friction — HB",
    "10_beam_cubicSpring_NM": "FE beam + cubic spring — NMA backbone",
    "analytical/duffing_backbone": "Duffing backbone (analytical, no MATLAB)",
}


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def load_fixture(name: str, fixtures_dir: Path | None = None) -> dict[str, Any]:
    """Load a fixture ``.npz`` file and return its contents as a plain dict.

    Parameters
    ----------
    name:
        Fixture name without the ``.npz`` extension, e.g. ``"01_Duffing"`` or
        ``"analytical/duffing_backbone"``.
    fixtures_dir:
        Override the default ``tests/fixtures/`` directory.  Useful in tests
        that need to point at a temporary directory.

    Returns
    -------
    dict[str, Any]
        Keys are the array names stored in the ``.npz`` file.  Every fixture
        guarantees ``omega``, ``amplitude``, ``phase``, ``stability``, and
        ``tolerance``.  Additional arrays may be present (see module docstring).

    Raises
    ------
    FileNotFoundError
        If the requested fixture file does not exist.
    """
    base_dir = fixtures_dir if fixtures_dir is not None else _FIXTURES_DIR
    path = base_dir / f"{name}.npz"
    if not path.exists():
        raise FileNotFoundError(
            f"Fixture not found: {path}\n"
            "Generate it with:  matlab -batch \"run('tools/generate_fixtures.m')\"\n"
            "or for the analytical fixture:  python tools/generate_fixtures.py --gen-analytical"
        )
    with np.load(str(path), allow_pickle=False) as npz:
        return {k: npz[k] for k in npz.files}


def list_fixtures(fixtures_dir: Path | None = None) -> list[dict[str, Any]]:
    """Return metadata for every ``.npz`` file found in *fixtures_dir*.

    Each entry is a dict with keys:

    - ``name``        — fixture name relative to fixtures_dir (no extension)
    - ``path``        — absolute Path to the file
    - ``present``     — bool
    - ``n_points``    — int or None  (length of ``omega`` array if present)
    - ``arrays``      — list[str]  (array names in the file)
    - ``description`` — str  (from ``_EXPECTED_FIXTURES`` catalogue, or "unknown")
    """
    base_dir = fixtures_dir if fixtures_dir is not None else _FIXTURES_DIR
    results: list[dict[str, Any]] = []

    # Walk expected list first so missing files are included in the output.
    seen: set[str] = set()
    for name, description in _EXPECTED_FIXTURES.items():
        path = base_dir / f"{name}.npz"
        present = path.exists()
        entry: dict[str, Any] = {
            "name": name,
            "path": path,
            "present": present,
            "n_points": None,
            "arrays": [],
            "description": description,
        }
        if present:
            with np.load(str(path), allow_pickle=False) as npz:
                entry["arrays"] = list(npz.files)
                if "omega" in npz.files:
                    entry["n_points"] = int(npz["omega"].shape[0])
        results.append(entry)
        seen.add(name)

    # Also surface any unexpected .npz files already in the directory.
    for npz_path in sorted(base_dir.rglob("*.npz")):
        rel = npz_path.relative_to(base_dir)
        name = str(rel.with_suffix(""))
        if name in seen:
            continue
        with np.load(str(npz_path), allow_pickle=False) as npz:
            arrays = list(npz.files)
            n_points = int(npz["omega"].shape[0]) if "omega" in npz.files else None
        results.append(
            {
                "name": name,
                "path": npz_path,
                "present": True,
                "n_points": n_points,
                "arrays": arrays,
                "description": "unknown",
            }
        )

    return results


def print_summary_table(fixtures_dir: Path | None = None) -> None:
    """Print a human-readable summary table of all fixtures to stdout."""
    rows = list_fixtures(fixtures_dir)
    col_name = max(len(r["name"]) for r in rows) + 2
    col_desc = 50
    header = (
        f"{'Fixture':<{col_name}} {'Present':<8} {'n_points':<10} "
        f"{'Arrays':<8} {'Description':<{col_desc}}"
    )
    separator = "-" * len(header)
    print(separator)
    print(header)
    print(separator)
    for r in rows:
        status = "YES" if r["present"] else "---"
        n_pts = str(r["n_points"]) if r["n_points"] is not None else "—"
        n_arr = str(len(r["arrays"])) if r["arrays"] else "—"
        desc = r["description"][:col_desc]
        print(
            f"{r['name']:<{col_name}} {status:<8} {n_pts:<10} {n_arr:<8} {desc:<{col_desc}}"
        )
    print(separator)
    present = sum(1 for r in rows if r["present"])
    print(f"\n{present}/{len(rows)} fixtures present in {_FIXTURES_DIR}\n")


def inspect_fixture(name: str, fixtures_dir: Path | None = None) -> None:
    """Print detailed information about a single fixture to stdout."""
    try:
        data = load_fixture(name, fixtures_dir)
    except FileNotFoundError as exc:
        print(f"ERROR: {exc}", file=sys.stderr)
        sys.exit(1)

    print(f"\nFixture: {name}")
    print(f"File:    {(_FIXTURES_DIR if fixtures_dir is None else fixtures_dir) / (name + '.npz')}")
    print(f"\n{'Array':<25} {'dtype':<10} {'shape':<20} {'min':>12} {'max':>12}")
    print("-" * 82)
    for key, val in data.items():
        arr = np.asarray(val)
        if arr.ndim == 0:
            # Scalar
            print(f"  {key:<23} {str(arr.dtype):<10} {'scalar':<20} {float(arr):>12.4g}")
        else:
            vmin = float(np.min(arr)) if np.issubdtype(arr.dtype, np.number) else "—"
            vmax = float(np.max(arr)) if np.issubdtype(arr.dtype, np.number) else "—"
            min_str = f"{vmin:>12.4g}" if isinstance(vmin, float) else f"{'—':>12}"
            max_str = f"{vmax:>12.4g}" if isinstance(vmax, float) else f"{'—':>12}"
            print(f"  {key:<23} {str(arr.dtype):<10} {str(arr.shape):<20} {min_str} {max_str}")
    print()


def generate_analytical_duffing_backbone(
    fixtures_dir: Path | None = None,
    n_points: int = 200,
    omega0: float = 1.0,
    k3: float = 1.0,
    a_min: float = 0.01,
    a_max: float = 2.0,
) -> Path:
    """Generate the analytical Duffing backbone curve and save as ``.npz``.

    The Duffing backbone relates modal amplitude *A* to natural frequency via:

        ω²(A) = ω₀² + (3/4) k₃ A²

    Reference: Krack & Gross (2019), §2.2, eq. (2.33) (single-harmonic
    approximation for conservative Duffing oscillator).

    Parameters
    ----------
    fixtures_dir:
        Directory under which ``analytical/`` will be created.  Defaults to
        ``tests/fixtures/``.
    n_points:
        Number of amplitude samples.  Defaults to 200.
    omega0:
        Linear natural frequency (rad/s).  Default 1.0.
    k3:
        Cubic stiffness coefficient.  Default 1.0.
    a_min:
        Minimum amplitude.  Default 0.01.
    a_max:
        Maximum amplitude.  Default 2.0.

    Returns
    -------
    Path
        Absolute path to the written ``.npz`` file.
    """
    base_dir = fixtures_dir if fixtures_dir is not None else _FIXTURES_DIR
    out_dir = base_dir / "analytical"
    out_dir.mkdir(parents=True, exist_ok=True)
    out_path = out_dir / "duffing_backbone.npz"

    amplitude = np.linspace(a_min, a_max, n_points)
    omega = np.sqrt(omega0**2 + 0.75 * k3 * amplitude**2)

    np.savez(str(out_path), amplitude=amplitude, omega=omega)
    return out_path


# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter,
    )
    group = parser.add_mutually_exclusive_group(required=True)
    group.add_argument(
        "--list",
        action="store_true",
        help="Print a summary table of all fixtures.",
    )
    group.add_argument(
        "--inspect",
        metavar="NAME",
        help="Inspect a single fixture (e.g. '01_Duffing').",
    )
    group.add_argument(
        "--gen-analytical",
        action="store_true",
        help="Generate (or regenerate) analytical/duffing_backbone.npz.",
    )
    parser.add_argument(
        "--fixtures-dir",
        metavar="DIR",
        type=Path,
        default=None,
        help="Override default tests/fixtures/ directory.",
    )
    return parser


def main(argv: list[str] | None = None) -> None:
    """Entry point for the CLI."""
    parser = _build_parser()
    args = parser.parse_args(argv)
    fdir: Path | None = args.fixtures_dir

    if args.list:
        print_summary_table(fdir)
    elif args.inspect:
        inspect_fixture(args.inspect, fdir)
    elif args.gen_analytical:
        path = generate_analytical_duffing_backbone(fdir)
        print(f"Written: {path}")


if __name__ == "__main__":
    main()

File: tools/generate_fixtures.m
% generate_fixtures.m
%
% MATLAB fixture generator for NLvib Python port validation.
%
% PURPOSE
%   Run each NLvib MATLAB example, capture the continuation results, normalise
%   them to the canonical fixture schema, and write one .npz file per example
%   to tests/fixtures/.  The resulting files are committed to the repository
%   and used as ground-truth by the Python QA agent.
%
% PREREQUISITES
%   1. MATLAB R2019b or newer  (Octave 7+ should also work)
%   2. NLvib MATLAB source present at  tools/NLvib_matlab/
%      (run  bash tools/fetch_matlab_source.sh  first)
%   3. scipy must NOT be required — the .npz format is written by a small
%      inline Python call using numpy only, invoked via system().
%      Alternatively the script writes .mat files and a companion Python
%      script converts them; see CONVERSION section below.
%
% USAGE
%   From the repo root:
%       matlab -batch "run('tools/generate_fixtures.m')"
%   or interactively:
%       >> run('tools/generate_fixtures.m')
%
% OUTPUT
%   tests/fixtures/01_Duffing.npz
%   tests/fixtures/02_twoDOFoscillator_cubicSpring.npz
%   tests/fixtures/03_twoDOFoscillator_unilateralSpring.npz
%   tests/fixtures/04_twoDOFoscillator_cubicSpring_NM.npz
%   tests/fixtures/05_twoDOFoscillator_tanhDryFriction_NM.npz
%   tests/fixtures/06_twoDOFoscillator_tanhDryFriction_FRF.npz
%   tests/fixtures/07_geometricNonlinearity.npz
%   tests/fixtures/08_multiDOF_multiNL.npz
%   tests/fixtures/09_beam_tanhFriction.npz
%   tests/fixtures/10_beam_cubicSpring_NM.npz
%
% REPRODUCIBILITY
%   All examples use the parameter values from the original NLvib distribution.
%   Do not change physical parameters here — the Python port is validated against
%   these exact values.  If parameters must change, update both this script and
%   the corresponding Python example, then regenerate all fixtures.
%
% SCHEMA (see also tests/fixtures/README.md)
%   Every fixture stores at minimum:
%     omega      (n_points,)  double   — frequency (rad/s)
%     amplitude  (n_points,)  double   — response amplitude at DOF 0, harmonic 1
%     phase      (n_points,)  double   — phase of fundamental harmonic (rad)
%     stability  (n_points,)  logical  — true = stable
%     tolerance  scalar       double   — validation tolerance (default 1e-6)
%
% ---------------------------------------------------------------------------

%% ---- Paths ----------------------------------------------------------------
script_dir  = fileparts(mfilename('fullpath'));      % tools/
repo_root   = fileparts(script_dir);                 % project root
nlvib_src   = fullfile(repo_root, 'tools', 'NLvib_matlab');
fixtures_dir = fullfile(repo_root, 'tests', 'fixtures');

if ~isfolder(nlvib_src)
    error(['NLvib MATLAB source not found at:\n  %s\n' ...
           'Run:  bash tools/fetch_matlab_source.sh'], nlvib_src);
end

% Add NLvib core to path (SRC subdirectory contains the solver functions)
addpath(genpath(fullfile(nlvib_src, 'SRC')));

%% ---- Helper: extract amplitude from harmonic coefficient matrix -----------
% Q is the raw HB solution matrix, shape (n_dof*(2*H+1)) x n_points.
% For DOF index d and harmonic h=1 the cosine/sine pair sits at rows
% (2*H+1)*d + 2*(h-1) + 1  and  (2*H+1)*d + 2*(h-1) + 2  (1-indexed, MATLAB).
% Amplitude = sqrt(cos^2 + sin^2).

function A = extract_amplitude(Q_branch, dof_idx, H)
    % EXTRACT_AMPLITUDE  Amplitude at a given DOF for the fundamental harmonic.
    %   Q_branch : matrix  (n_dof*(2*H+1)) x n_points
    %   dof_idx  : 0-based DOF index
    %   H        : number of harmonics
    n_coeff_per_dof = 2*H + 1;
    row_cos = n_coeff_per_dof * dof_idx + 2;   % 1-based; row 1 is constant term
    row_sin = n_coeff_per_dof * dof_idx + 3;
    A = sqrt(Q_branch(row_cos, :).^2 + Q_branch(row_sin, :).^2);
    A = A(:);  % column vector
end

function phi = extract_phase(Q_branch, dof_idx, H)
    % EXTRACT_PHASE  Phase angle at a given DOF for the fundamental harmonic.
    n_coeff_per_dof = 2*H + 1;
    row_cos = n_coeff_per_dof * dof_idx + 2;
    row_sin = n_coeff_per_dof * dof_idx + 3;
    phi = atan2(Q_branch(row_sin, :), Q_branch(row_cos, :));
    phi = phi(:);
end

%% ---- Helper: write .mat then convert to .npz via Python -------------------

function write_fixture(fixture_name, data_struct, fixtures_dir, repo_root)
    % WRITE_FIXTURE  Save a struct to .mat and convert to .npz using Python.
    %   fixture_name : string without extension, e.g. '01_Duffing'
    %   data_struct  : MATLAB struct with fields matching the npz schema
    %   fixtures_dir : absolute path to tests/fixtures/
    %   repo_root    : absolute path to repo root
    mat_path = fullfile(fixtures_dir, [fixture_name '.mat']);
    npz_path = fullfile(fixtures_dir, [fixture_name '.npz']);

    % Save to .mat (v7 is compatible with scipy.io.loadmat)
    save(mat_path, '-struct', 'data_struct', '-v7');
    fprintf('  Written: %s\n', mat_path);

    % Convert to .npz using a one-liner Python call
    py_convert = sprintf([ ...
        'import numpy as np, scipy.io as sio; ' ...
        'm = sio.loadmat(r"%s"); ' ...
        'keys = [k for k in m if not k.startswith("_")]; ' ...
        'arrays = {}; ' ...
        '[arrays.update({k: m[k].squeeze()}) for k in keys]; ' ...
        'np.savez(r"%s", **arrays); ' ...
        'print("Converted:", r"%s")' ...
        ], mat_path, npz_path, npz_path);

    ret = system(sprintf('python3 -c "%s"', py_convert));
    if ret == 0
        delete(mat_path);   % remove intermediate .mat file
        fprintf('  Fixture:  %s\n', npz_path);
    else
        warning('Python conversion failed for %s — .mat file retained.', fixture_name);
    end
end

%% ===========================================================================
%% EXAMPLE 01 — Duffing oscillator  (T-15)
%% ===========================================================================
% Reference: NLvib EXAMPLES/01_Duffing/
% System:    m*xdd + d*xd + k*x + k3*x^3 = f*cos(Omega*t)
% Parameters from NLvib default: m=1, d=0.01, k=1, k3=1, f=0.1
% Methods: Harmonic Balance (H=7) + Shooting
% ---------------------------------------------------------------------------
fprintf('\n=== Example 01: Duffing oscillator ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '01_Duffing');
    orig_dir = cd(ex_dir);

    % Run the example script — it must define Om_HB, Q_HB, Om_shoot, Q_shoot,
    % and S_HB (stability from Floquet), S_shoot in the workspace.
    run('Duffing.m');

    cd(orig_dir);

    H = 7;   % number of harmonics used in the Duffing example

    d01.omega         = Om_HB(:);
    d01.amplitude     = extract_amplitude(Q_HB, 0, H);
    d01.phase         = extract_phase(Q_HB, 0, H);
    d01.stability     = S_HB(:) > 0;   % positive real part = unstable in NLvib convention
    d01.tolerance     = 1e-6;

    % Shooting overlay (additional arrays, not part of minimal schema)
    if exist('Om_shoot', 'var') && exist('Q_shoot', 'var')
        d01.omega_shoot     = Om_shoot(:);
        d01.amplitude_shoot = extract_amplitude(Q_shoot, 0, H);
        if exist('S_shoot', 'var')
            d01.stability_shoot = S_shoot(:) > 0;
        end
    end

    % Full harmonic matrix for residual-level validation
    d01.Q_harmonics = Q_HB;   % (n_dof*(2H+1)) x n_points

    write_fixture('01_Duffing', d01, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 02 — 2-DOF oscillator with cubic spring  (T-16)
%% ===========================================================================
% Reference: NLvib EXAMPLES/02_twoDOFoscillator_cubicSpring/
% System:    2-DOF chain; cubic spring between DOF 1 and wall
% Method:    Harmonic Balance
% ---------------------------------------------------------------------------
fprintf('\n=== Example 02: 2-DOF cubic spring ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '02_twoDOFoscillator_cubicSpring');
    orig_dir = cd(ex_dir);
    run('twoDOFoscillator_cubicSpring.m');
    cd(orig_dir);

    H = 7;
    d02.omega         = Om(:);
    d02.amplitude     = extract_amplitude(Q, 0, H);
    d02.phase         = extract_phase(Q, 0, H);
    d02.amplitude_dof2 = extract_amplitude(Q, 1, H);
    d02.phase_dof2    = extract_phase(Q, 1, H);
    d02.stability     = S(:) > 0;
    d02.tolerance     = 1e-6;
    d02.Q_harmonics   = Q;

    write_fixture('02_twoDOFoscillator_cubicSpring', d02, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 03 — 2-DOF oscillator with unilateral spring  (T-17)
%% ===========================================================================
% Reference: NLvib EXAMPLES/03_twoDOFoscillator_unilateralSpring/
% System:    2-DOF chain; unilateral contact spring on DOF 2
% Method:    Harmonic Balance (higher harmonics needed for impact dynamics)
% Note:      Non-smooth nonlinearity — tolerance relaxed to 1e-4
% ---------------------------------------------------------------------------
fprintf('\n=== Example 03: 2-DOF unilateral spring ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '03_twoDOFoscillator_unilateralSpring');
    orig_dir = cd(ex_dir);
    run('twoDOFoscillator_unilateralSpring.m');
    cd(orig_dir);

    H = 7;
    d03.omega         = Om(:);
    d03.amplitude     = extract_amplitude(Q, 0, H);
    d03.phase         = extract_phase(Q, 0, H);
    d03.amplitude_dof2 = extract_amplitude(Q, 1, H);
    d03.phase_dof2    = extract_phase(Q, 1, H);
    d03.stability     = S(:) > 0;
    d03.tolerance     = 1e-4;   % relaxed: non-smooth NL
    d03.Q_harmonics   = Q;

    write_fixture('03_twoDOFoscillator_unilateralSpring', d03, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 04 — 2-DOF cubic spring, Nonlinear Modal Analysis  (T-18)
%% ===========================================================================
% Reference: NLvib EXAMPLES/04_twoDOFoscillator_cubicSpring_NM/
% System:    Same as Example 02
% Method:    Nonlinear Normal Modes (backbone curve via NMA)
% ---------------------------------------------------------------------------
fprintf('\n=== Example 04: 2-DOF cubic spring, NMA backbone ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '04_twoDOFoscillator_cubicSpring_NM');
    orig_dir = cd(ex_dir);
    run('twoDOFoscillator_cubicSpring_NM.m');
    cd(orig_dir);

    H = 7;
    % NMA: Om is the backbone frequency (natural, not excitation)
    d04.omega         = Om(:);
    d04.amplitude     = extract_amplitude(Q, 0, H);
    d04.phase         = extract_phase(Q, 0, H);
    d04.amplitude_dof2 = extract_amplitude(Q, 1, H);
    d04.phase_dof2    = extract_phase(Q, 1, H);
    d04.stability     = ones(length(Om), 1) > 0;   % NMA backbone: stability not tracked
    d04.tolerance     = 1e-6;
    d04.Q_harmonics   = Q;

    write_fixture('04_twoDOFoscillator_cubicSpring_NM', d04, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 05 — 2-DOF oscillator with tanh dry friction, NMA  (T-18)
%% ===========================================================================
% Reference: NLvib EXAMPLES/05_twoDOFoscillator_tanhDryFriction_NM/
% System:    2-DOF chain; tanh-regularised Coulomb friction on DOF 2
% Method:    NMA backbone curve
% ---------------------------------------------------------------------------
fprintf('\n=== Example 05: 2-DOF tanh friction, NMA backbone ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '05_twoDOFoscillator_tanhDryFriction_NM');
    orig_dir = cd(ex_dir);
    run('twoDOFoscillator_tanhDryFriction_NM.m');
    cd(orig_dir);

    H = 7;
    d05.omega         = Om(:);
    d05.amplitude     = extract_amplitude(Q, 0, H);
    d05.phase         = extract_phase(Q, 0, H);
    d05.amplitude_dof2 = extract_amplitude(Q, 1, H);
    d05.phase_dof2    = extract_phase(Q, 1, H);
    d05.stability     = ones(length(Om), 1) > 0;
    d05.tolerance     = 1e-4;   % relaxed: tanh smoothing introduces small discretisation error
    d05.Q_harmonics   = Q;

    write_fixture('05_twoDOFoscillator_tanhDryFriction_NM', d05, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 06 — 2-DOF tanh friction, FRF  (T-18, forced response)
%% ===========================================================================
% Reference: same example folder as 05 but with external forcing applied
% Method:    Harmonic Balance FRF sweep
% ---------------------------------------------------------------------------
fprintf('\n=== Example 06: 2-DOF tanh friction, HB FRF ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '05_twoDOFoscillator_tanhDryFriction_NM');
    orig_dir = cd(ex_dir);

    % The NLvib example script typically computes both NMA and FRF sections.
    % Variable names for FRF branch are Om_FRF / Q_FRF in newer NLvib versions.
    % Adjust variable names if the example uses different identifiers.
    run('twoDOFoscillator_tanhDryFriction_NM.m');
    cd(orig_dir);

    H = 7;
    % Try to use Om_FRF / Q_FRF; fall back to Om / Q if not present.
    if exist('Om_FRF', 'var') && exist('Q_FRF', 'var')
        om_frf = Om_FRF;
        q_frf  = Q_FRF;
    else
        om_frf = Om;
        q_frf  = Q;
    end

    d06.omega         = om_frf(:);
    d06.amplitude     = extract_amplitude(q_frf, 0, H);
    d06.phase         = extract_phase(q_frf, 0, H);
    d06.amplitude_dof2 = extract_amplitude(q_frf, 1, H);
    d06.phase_dof2    = extract_phase(q_frf, 1, H);
    d06.stability     = S(:) > 0;
    d06.tolerance     = 1e-4;
    d06.Q_harmonics   = q_frf;

    write_fixture('06_twoDOFoscillator_tanhDryFriction_FRF', d06, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 07 — Multi-DOF geometric nonlinearity  (T-19)
%% ===========================================================================
% Reference: NLvib EXAMPLES/06_geometricNonlinearity/  (check folder name)
% System:    Multi-DOF chain with geometric (displacement-squared) nonlinearity
% Method:    Harmonic Balance
% ---------------------------------------------------------------------------
fprintf('\n=== Example 07: Multi-DOF geometric nonlinearity ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '06_geometricNonlinearity');
    if ~isfolder(ex_dir)
        % Some NLvib distributions name this folder differently
        ex_dir = fullfile(nlvib_src, 'EXAMPLES', '06_multiDOFoscillator_geometricNL');
    end
    orig_dir = cd(ex_dir);
    % Run whichever .m file is present in this folder
    mfiles = dir('*.m');
    run(mfiles(1).name);
    cd(orig_dir);

    H = 7;
    d07.omega     = Om(:);
    d07.amplitude = extract_amplitude(Q, 0, H);
    d07.phase     = extract_phase(Q, 0, H);
    d07.stability = S(:) > 0;
    d07.tolerance = 1e-6;
    d07.Q_harmonics = Q;

    write_fixture('07_geometricNonlinearity', d07, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 08 — Multi-DOF oscillator with multiple NL elements  (T-20)
%% ===========================================================================
% Reference: NLvib EXAMPLES/06_multiDOFoscillator/ (check exact folder name)
% System:    Chain of oscillators with multiple different nonlinear attachments
% Method:    Harmonic Balance
% ---------------------------------------------------------------------------
fprintf('\n=== Example 08: Multi-DOF multi-NL ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '06_multiDOFoscillator');
    if ~isfolder(ex_dir)
        ex_dir = fullfile(nlvib_src, 'EXAMPLES', '06_multiDOF_multiNL');
    end
    orig_dir = cd(ex_dir);
    mfiles = dir('*.m');
    run(mfiles(1).name);
    cd(orig_dir);

    H = 7;
    d08.omega     = Om(:);
    d08.amplitude = extract_amplitude(Q, 0, H);
    d08.phase     = extract_phase(Q, 0, H);
    d08.stability = S(:) > 0;
    d08.tolerance = 1e-6;
    d08.Q_harmonics = Q;

    write_fixture('08_multiDOF_multiNL', d08, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 09 — FE beam with tanh friction  (T-21)
%% ===========================================================================
% Reference: NLvib EXAMPLES/07_beam_tanhFriction/
% System:    Euler-Bernoulli FE beam (n_elements=10) with tanh friction attachment
% Method:    Harmonic Balance
% Note:      Larger system — may take 10–20 minutes on a laptop
% ---------------------------------------------------------------------------
fprintf('\n=== Example 09: FE beam + tanh friction ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '07_beam_tanhFriction');
    orig_dir = cd(ex_dir);
    mfiles = dir('*.m');
    run(mfiles(1).name);
    cd(orig_dir);

    H = 7;
    d09.omega     = Om(:);
    d09.amplitude = extract_amplitude(Q, 0, H);
    d09.phase     = extract_phase(Q, 0, H);
    d09.stability = S(:) > 0;
    d09.tolerance = 1e-4;   % relaxed: non-smooth + large FE system
    d09.Q_harmonics = Q;

    write_fixture('09_beam_tanhFriction', d09, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% EXAMPLE 10 — FE beam with cubic spring, NMA  (T-22)
%% ===========================================================================
% Reference: NLvib EXAMPLES/08_beam_cubicSpring_NM/
% System:    Euler-Bernoulli FE beam (n_elements=10) with cubic spring midpoint
% Method:    NMA backbone curve
% Note:      Large system; CMS reduction recommended before NMA
% ---------------------------------------------------------------------------
fprintf('\n=== Example 10: FE beam + cubic spring, NMA backbone ===\n');
try
    ex_dir = fullfile(nlvib_src, 'EXAMPLES', '08_beam_cubicSpring_NM');
    orig_dir = cd(ex_dir);
    mfiles = dir('*.m');
    run(mfiles(1).name);
    cd(orig_dir);

    H = 7;
    d10.omega     = Om(:);
    d10.amplitude = extract_amplitude(Q, 0, H);
    d10.phase     = extract_phase(Q, 0, H);
    d10.stability = ones(length(Om), 1) > 0;   % NMA: stability not tracked
    d10.tolerance = 1e-6;
    d10.Q_harmonics = Q;

    write_fixture('10_beam_cubicSpring_NM', d10, fixtures_dir, repo_root);
    fprintf('  OK\n');
catch ME
    fprintf('  FAILED: %s\n', ME.message);
end

%% ===========================================================================
%% Summary
%% ===========================================================================
fprintf('\n========================================\n');
fprintf('Fixture generation complete.\n');
fprintf('Output directory: %s\n', fixtures_dir);
fprintf('Run  python tools/generate_fixtures.py --list  to inspect.\n');
fprintf('========================================\n');
